examples/ppo/main.py

```diff
@@ -0,0 +1,295 @@
+    vf_coeff: weighs value function loss in total loss
+    entropy_coeff: weighs entropy bonus in the total loss
+  Returns:
+    optimizer: new optimizer after the parameters update
+    loss: loss summed over training steps
+    grad_norm: gradient norm from last step (summed over parameters)
+  """
+  def loss_fn(model, minibatch, clip_param, vf_coeff, entropy_coeff):
+    states, actions, old_log_probs, returns, advantages = minibatch
+    shapes = list(map(lambda x : x.shape, minibatch))
+    assert(shapes[0] == (BATCH_SIZE, 84, 84, 4))
+    assert(all(s == (BATCH_SIZE,) for s in shapes[1:]))
+    log_probs, values = model(states)
+    values = values[:, 0] # convert shapes: (batch, 1) to (batch, )
+    probs = jnp.exp(log_probs)
+    entropy = jnp.sum(-probs*log_probs, axis=1).mean()
+    log_probs_act_taken = jax.vmap(lambda lp, a: lp[a])(log_probs, actions)
+    ratios = jnp.exp(log_probs_act_taken - old_log_probs)
+    # adv. normalization (following the OpenAI baselines)
+    advantages = (advantages - advantages.mean()) / (advantages.std() + 1e-8)
+    PG_loss = ratios * advantages
+    clipped_loss = advantages * jax.lax.clamp(1. - clip_param, ratios,
+                                               1. + clip_param)
+    assert(PG_loss.shape == clipped_loss.shape)
+    PPO_loss = -jnp.mean(jnp.minimum(PG_loss, clipped_loss), axis=0)
+    assert(values.shape == returns.shape)
+    value_loss = jnp.mean(jnp.square(returns - values), axis=0)
+    return PPO_loss + vf_coeff*value_loss - entropy_coeff*entropy
+
+  batch_size = BATCH_SIZE
+  iterations = trn_data[0].shape[0] // batch_size
+  trn_data = jax.tree_map(
+    lambda x: x.reshape((iterations, batch_size) + x.shape[1:]), trn_data)
+  loss = 0.
+  for batch in zip(*trn_data):
+    grad_fn = jax.value_and_grad(loss_fn)
+    l, grad = grad_fn(optimizer.target, batch, clip_param, vf_coeff,
+    entropy_coeff)
+    loss += l
+    optimizer = optimizer.apply_gradient(grad)
+    grad_norm = sum(jnp.square(g).sum() for g in jax.tree_leaves(grad))
+  return optimizer, loss, grad_norm
+
+
+def thread_inference(
+  q1 : Queue,
+  q2: Queue,
+  simulators : List[RemoteSimulator],
+  steps_per_actor : int):
+  """Worker function for a separate thread used for inference and running
+  the simulators in order to maximize the GPU/TPU usage. Runs
+  `steps_per_actor` time steps of the game for each of the `simulators`.
+  """
+
+  while(True):
+    optimizer, step = q1.get()
+    all_experience = []
+    for _ in range(steps_per_actor + 1): # +1 to get one more value
+                                         # needed for GAE
+      states = []
+      for sim in simulators:
+        state = sim.conn.recv()
+        states.append(state)
+      states = onp.concatenate(states, axis=0)
+
+      # perform inference
+      # policy_optimizer, step = q1.get()
+      log_probs, values = policy_action(optimizer.target, states)
```

Comment:
log_probs, _ = policy_action(optimizer.target, states)
log_probs = jax.device_get(log_probs)

Reply:
Actually it should just be:
```
log_probs, values = jax.device_get((log_probs, values))
```

Reply:
https://github.com/google/flax/pull/470/commits/f88e45b00df0199db998eb784c42149854c14c1a